### ISSUE TITLE
CADE: Anomaly detection, "one-class" classifier

### ISSUE THREAD

**Issue by choshin84** [choshin84/learning_memo#8]
# Tweet summary
1. Anomaly detection
- All real data set + fake data set with uniform distributed features
(uniform distribution can be easily distinguish vs. original distribution esp. far edge)
- Build model to separate real/fake, infere probability on all data set
- Rank by fake probability to see anomality
2. "One-class" classification
- Majority class data set + fake data set with normal / shuffle distributed features
- Builid model to separate real/fake
- Test on real data with 2-class

# Useful link
http://darrkj.github.io/blog/2014/may102014/

**Comment by choshin84** [comment]
"One-class" classification may need to check if it can distinguish another class before applying to entire data set.

**Comment by choshin84** [comment]
CADE class
```python
import pandas as pd 
import numpy as np 
from numpy.random import choice
import sklearn
import sklearn.datasets
import sklearn.ensemble
import sklearn.metrics


class CADE():
    # Define the core functions to form synthetic rows from real rows.
    def synthetic_generator_uniform(df):
        NOMINAL_DTYPES = ['object']
        def synth(feature):
            if feature.dtype.name in NOMINAL_DTYPES:
                synth = sklearn.utils.resample(
                    feature, n_samples=feature.shape[0]
                ).values
            else:
                synth = np.random.uniform(
                    1.0 * feature.min(), 1.0 * feature.max(), feature.shape[0]
                )
            return synth

        return pd.DataFrame({
            colname: synth(feature) for (colname, feature) in df.iteritems()
        })[df.columns]

    def synthetic_generator_normal(df):
        NOMINAL_DTYPES = ['object']
        def synth(feature):
            if feature.dtype.name in NOMINAL_DTYPES:
                synth = sklearn.utils.shuffle(feature).values
            else:
                synth = np.random.normal(
                    np.mean(feature), np.std(feature, ddof=1), feature.shape[0]
                )
            return synth

        return pd.DataFrame({
            colname: synth(feature) for (colname, feature) in df.iteritems()
        })[df.columns]

    def synthetic_generator_shuffled(df):
        return pd.DataFrame({
            colname: sklearn.utils.shuffle(feature).values for (colname, feature) in df.iteritems()
        })[df.columns]
```
